### ISSUE TITLE
`on_attach` "global hook" not working, but individual lspconfig setup `on_attach` is working

### ISSUE THREAD

**Issue by suvash** [VonHeikemen/lsp-zero.nvim#338]
As demonstrated in the docs, and this blog post ( https://dev.to/vonheikemen/make-lsp-zeronvim-coexists-with-other-plugins-instead-of-controlling-them-2i80 ), when I use a minimal config as below.
```lua
local lsp_zero = require('lsp-zero')

lsp_zero.on_attach(function(client, bufnr)
  lsp_zero.default_keymaps({buffer = bufnr})
  print(client.name .. ': Hello there.')
end)

require('lspconfig').tsserver.setup({
  on_attach = function(client, bufnr)
    print('tsserver: How are you today?')
  end
})
```

the `lsp_zero.on_attach` does not seem to be working, (or perhaps it's getting overridden somewhere else), but the individual lsp server on_attach seems to be working.

I can confirm that the lsp servers are working, but I can't get the default keymaps (or anything) setup in the global hook. At the moment, I'm repeating the keymaps in each individual server config., but it would be nice to figure out what actually is going wrong the first place.

Could you provide any hints to how I would go debugging this ?

**Comment by VonHeikemen** [comment]
First, you can check that lsp-zero's initial setup was executed. A global variable called `loaded_lsp_zero` should exists. Execute this command.

```vim
:echo g:loaded_lsp_zero
```

That should print the value `1`.

Then you can check if the autocommand `LspAttach` was created.

```vim
:autocmd LspAttach
```

To verify that the autocommand was executed in the current buffer you can check if the command `LspZeroFormat` exists.

```vim
:command LspZeroFormat
```

**Comment by suvash** [comment]
Thanks for the quick reply. 

When checking if lsp-zero's initial setup was executed (via `:echo g:loaded_lsp_zero`), it's undefined.
<img width="376" alt="Screen Shot 2023-11-05 at 22 01 36" src="https://github.com/VonHeikemen/lsp-zero.nvim/assets/144952/b077cb35-2798-4dbe-8330-1fb1833f93c0">

However, when running `:LspInfo` right away, I see that 2 clients are attached to the buffer. Perhaps I'm missing something about how lsp-zero is initialised, (or how lsp servers are supposed to be initialised)
<img width="575" alt="Screen Shot 2023-11-05 at 22 05 45" src="https://github.com/VonHeikemen/lsp-zero.nvim/assets/144952/e54b84a0-fe4f-4d1e-9162-d8308a2c716e">

The next check `:autocmd LspAttach` does return some result, though a bit empty perhaps.
<img width="406" alt="Screen Shot 2023-11-05 at 22 08 48" src="https://github.com/VonHeikemen/lsp-zero.nvim/assets/144952/b91833e0-043f-43b7-9ce5-fac405d762d4">

The final suggestion does not yield any result.
<img width="283" alt="Screen Shot 2023-11-05 at 22 12 18" src="https://github.com/VonHeikemen/lsp-zero.nvim/assets/144952/607ccba3-4a52-4042-8985-c4da8f0457ee">

Where should I go from here ?

For some more context, I don't use mason & mason-lspconfig. The neovim plugins are pre-installed directly via nixpkgs(using home manager), and the same thing with the language server binaries.
I can confirm that the server binaries are on the path, and I've had no issues so far with any other plugins, so I'm sort of trusting that the installations are not messed up somehow.

One more question, when I manually trigger the `:LspRestart`, should I be expecting the "global" hook to be re-triggered as well, or is that only a one time thing per buffer ?

**Comment by VonHeikemen** [comment]
The problem here is the initial setup of lsp-zero is not being executed. Neovim should execute [this file](https://github.com/VonHeikemen/lsp-zero.nvim/blob/v3.x/plugin/lsp-zero.lua) automatically, but is not happening. You need the autocommand to exists so the "global hook" (and other features) can work.

The ideal solution to this would be figuring out why Neovim is skipping lsp-zero's plugin script, and then fix that.

The dirty hack that you can do to try to make it work is to load the module `lsp-zero.setup` manually.

```lua
require('lsp-zero.setup')

local lsp_zero = require('lsp-zero')
----
--- the rest of your config goes here...
---
```

**Comment by suvash** [comment]
Thanks for the quick reply again.

I'll try your suggestions and follow up with the results. I'll also try to ensure that I'm running the 3.0 version of the plugin and see if it's still the case.

I'm on the road today, but will have some answers in a day or two. 🙌🏽

**Comment by VonHeikemen** [comment]
I have a question:

>  The neovim plugins are pre-installed directly via nixpkgs(using home manager), and the same thing with the language server binaries.

Are you using [this package](https://search.nixos.org/packages?channel=23.05&from=0&size=50&sort=relevance&type=packages&query=lsp-zero) from nixpkgs?

That is pinned to an old commit from the main branch. Versions in lsp-zero have their branch (I still push bug fixes to old versions).

Whoever packages lsp-zero for nixpkgs should update the commit hash to something in the `v3.x` branch.

**Comment by suvash** [comment]
> 

> Are you using [this package](https://search.nixos.org/packages?channel=23.05&from=0&size=50&sort=relevance&type=packages&query=lsp-zero) from nixpkgs?

Yes, that's the one. But, I'll try to get the plugin installed from the 3.x branch today and hopefully all will start working well.

**Comment by suvash** [comment]
I just installed the plugin directly from the 3.x branch (using flake inputs to build the vimplugin via nix), and I can now confirm that all the debugging checks above works. And the global hook on attach also runs, given the print statement finally being emitted.
I'll do some more test and report back if something's not working as expected. But, for the time being, I'll close this issue. I'll also try to send a PR to update the nixpkgs upstream to use the 3.x branch.
Thanks for all the help. 🙌🏽

**Comment by suvash** [comment]
@VonHeikemen Could you chime in the discussion above and let the maintainers in nixpkgs officially know that `3.x` is the expected default branch to be packaged ?

**Comment by steventsvetkov** [comment]
I encountered a very similar issue while installing the plugin though lazy.nvim. This is the config

```lua
{
    "VonHeikemen/lsp-zero.nvim",
    branch = "v3.x",
    dependencies = {
        -- LSP Support
        { "neovim/nvim-lspconfig" },
        { "williamboman/mason.nvim" },
        { "williamboman/mason-lspconfig.nvim" },

        -- Autocompletion
        { "hrsh7th/nvim-cmp" },
        { "hrsh7th/cmp-nvim-lsp" },
        { "hrsh7th/cmp-buffer" },
        { "hrsh7th/cmp-path" },
        { "saadparwaiz1/cmp_luasnip" },
        { "hrsh7th/cmp-nvim-lua" },
        { "zbirenbaum/copilot.lua" },
        { "zbirenbaum/copilot-cmp" },

        -- Snippets
        { "L3MON4D3/LuaSnip" },
        { "rafamadriz/friendly-snippets" },
    },
    config = function()
        require("plugins.lsp")
    end,
})
```

`plugins.lsp` holds Prime's config from the docs.

Regardless if I install it with this configuration, or just the plugin without dependencies though `{'VonHeikemen/lsp-zero.nvim', branch = 'v3.x'},`, or even if I skip the branch and only leave `{'VonHeikemen/lsp-zero.nvim'},`, lazy.nvim for some reason pulls the repo and instead of checking out the `v3.x` branch, it checks out the c51a910 commit from a few weeks ago, which results in `HEAD detached at c51a910`.

The plugin at that point in time does not appear to startup properly (or maybe I'm missing something) and the debugs above fail. The "solution" I found was to manually go to `~/.local/share/nvim/lazy/lsp-zero.nvim` and call `git checkout v3.x`. Now everything works as expected.
